tool/tsh/db.go

```diff
@@ -0,0 +1,223 @@
+		servers, err = tc.ListDatabaseServers(cf.Context)
+		return trace.Wrap(err)
+	})
+	if err != nil {
+		utils.FatalError(err)
+	}
+	sort.Slice(servers, func(i, j int) bool {
+		return servers[i].GetName() < servers[j].GetName()
+	})
+	// Retrieve profile to be able to show which databases user is logged into.
+	profile, err := client.StatusCurrent("", cf.Proxy)
+	if err != nil {
+		utils.FatalError(err)
+	}
+	showDatabases(servers, profile.Databases, cf.Verbose)
+}
+
+// onDatabaseLogin handles "tsh db login" command.
+func onDatabaseLogin(cf *CLIConf) {
+	tc, err := makeClient(cf, false)
+	if err != nil {
+		utils.FatalError(err)
+	}
+	var servers []services.DatabaseServer
+	err = client.RetryWithRelogin(cf.Context, tc, func() error {
+		servers, err = tc.ListDatabaseServersFor(cf.Context, cf.DatabaseService)
+		return trace.Wrap(err)
+	})
+	if err != nil {
+		utils.FatalError(err)
+	}
+	if len(servers) == 0 {
+		utils.FatalError(trace.NotFound(
+			"database %q not found, use 'tsh db ls' to see registered databases", cf.DatabaseService))
+	}
+	err = databaseLogin(cf, tc, tlsca.RouteToDatabase{
+		ServiceName: cf.DatabaseService,
+		Protocol:    servers[0].GetProtocol(),
+		Username:    cf.DatabaseUser,
+		Database:    cf.DatabaseName,
+	}, false)
+	if err != nil {
+		utils.FatalError(err)
+	}
+}
+
+func databaseLogin(cf *CLIConf, tc *client.TeleportClient, db tlsca.RouteToDatabase, quiet bool) error {
+	log.Debugf("Fetching database access certificate for %s on cluster %v.", db, tc.SiteName)
+	profile, err := client.StatusCurrent("", cf.Proxy)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	err = tc.ReissueUserCerts(cf.Context, client.ReissueParams{
+		RouteToCluster: tc.SiteName,
+		RouteToDatabase: proto.RouteToDatabase{
+			ServiceName: db.ServiceName,
+			Protocol:    db.Protocol,
+			Username:    db.Username,
+			Database:    db.Database,
+		},
+		AccessRequests: profile.ActiveRequests.AccessRequests,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// Refresh the profile.
+	profile, err = client.StatusCurrent("", cf.Proxy)
+	if err != nil {
+		utils.FatalError(err)
+	}
+	// Perform database-specific actions such as updating Postgres
+	// connection service file.
+	switch db.Protocol {
+	case defaults.ProtocolPostgres:
+		err := pgservicefile.Add(tc.SiteName, db.ServiceName, db.Username, db.Database, *profile, quiet)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	return nil
+}
+
+// fetchDatabaseCreds is called as a part of tsh login to refresh database
+// access certificates for databases the current profile is logged into.
+func fetchDatabaseCreds(cf *CLIConf, tc *client.TeleportClient) error {
+	profile, err := client.StatusCurrent("", cf.Proxy)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	for _, db := range profile.Databases {
+		if err := databaseLogin(cf, tc, db, true); err != nil {
+			log.WithError(err).Errorf("Failed to fetch database access certificate for %s.", db)
```

Comment:
Why are these not error conditions?

Reply:
This is for the case when you execute `tsh login` while still being logged into some databases (say, the next day). This case makes sure that login doesn't fail if any of the databases have disappeared for example (which can happen) - in that case their login info will just be cleared.